torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -374,6 +374,361 @@ def collect_bw_donated_buffer_idxs(
+            )
+            if node.args[1].startswith("___forward"):
+                # Insert the new_fw_hop_gm. This is straightforward. Get the
+                # new_fw_hop_gm, insert the hop_gm as a get_attr node, and then
+                # add a call_function node. Additionally, also use getitem
+                # call_functions to collect the saved_tensor nodes
+
+                new_fw_hop_gm = new_hop_graphs[identifier].new_fw_hop_gm
+                assert new_fw_hop_gm is not None
+                new_fw_mod_attr_name = add_new_hop_gm(new_fw_hop_gm, node.args[1])
+                new_fw_mod_attr = new_graph.get_attr(new_fw_mod_attr_name)
+
+                old_operands = node.args[2]
+                new_operands = [env[n] for n in old_operands]
+                env[node] = new_graph.call_function(
+                    the_function=invoke_subgraph,
+                    args=(
+                        new_fw_mod_attr,
+                        new_fw_mod_attr_name,
+                        tuple(new_operands),
+                    ),
+                )
+                propagate_meta_info(new_fw_hop_gm, env[node], node)
+
+                old_num_fw_outputs = new_hop_graphs[identifier].old_num_fw_outputs
+                new_num_sym_nodes = new_hop_graphs[identifier].new_num_sym_nodes
+                new_num_saved_nodes = new_hop_graphs[identifier].new_num_saved_nodes
+                assert old_num_fw_outputs is not None
+                assert new_num_sym_nodes is not None
+                assert new_num_saved_nodes is not None
+
+                extra_fw_outputs = []
+                # new_hop_fw_gm output signature is (*fw_outs, *saved_tensors)
+                # old_num_fw_outputs = len(fw_outs)
+                # new_num_fw_outputs = len(*fw_outs, *saved_tensors, *sym_nodes)
+                total_outputs = (
+                    old_num_fw_outputs + new_num_saved_nodes + new_num_sym_nodes
+                )
+                for fw_out_idx in range(old_num_fw_outputs, total_outputs):
+                    saved_tensor_node = new_graph.call_function(
+                        the_function=operator.getitem, args=(env[node], fw_out_idx)
+                    )
+                    saved_tensor_node.meta = copy.copy(env[node].meta)
+                    saved_tensor_node.meta["val"] = env[node].meta["val"][fw_out_idx]
+                    extra_fw_outputs.append(saved_tensor_node)
+
+                # Save the saved_tensors info for the fw_node. This will be used
+                # to form the inputs for the backward hop.
+                old_fw_hop_nodes_stack.append(node)
+                fw_node_to_saved_tensors_map[node] = (identifier, extra_fw_outputs)
+            elif node.args[1].startswith("___backward"):
+                # Get the saved_tensors from the forward graph and find the new
+                # tangents, and replace the old bw hop with the new bw hop.
```

Comment:
# Now, lets look at the backward hop graph. Old signature
    #   inputs - (*primals, *tangents)
    #   outputs - (*grad_outs, *fw_outs)
    # New signature
    #   inputs - (*saved_tensors, *tangents) -- Different
    #   outputs - (*grad_outs)  -- Different